core/src/test/java/org/elasticsearch/common/geo/GeoWKTShapeParserTests.java

```diff
@@ -0,0 +1,182 @@
+        Coordinate c = new Coordinate(p.lon(), p.lat());
+        Point expected = GEOMETRY_FACTORY.createPoint(c);
+        assertExpected(new JtsPoint(expected, SPATIAL_CONTEXT), new PointBuilder().coordinate(c));
+    }
+
+    @Override
+    public void testParseMultiPoint() throws IOException {
+        int numPoints = randomIntBetween(2, 100);
+        List<Coordinate> coordinates = new ArrayList<>(numPoints);
+        Shape[] shapes = new Shape[numPoints];
+        GeoPoint p = new GeoPoint();
+        for (int i = 0; i < numPoints; ++i) {
+            p.reset(GeoTestUtil.nextLatitude(), GeoTestUtil.nextLongitude());
+            coordinates.add(new Coordinate(p.lon(), p.lat()));
+            shapes[i] = SPATIAL_CONTEXT.makePoint(p.lon(), p.lat());
+        }
+        ShapeCollection expected = shapeCollection(shapes);
+        assertExpected(expected, new MultiPointBuilder(coordinates));
+    }
+
+    private List<Coordinate> randomLineStringCoords() {
+        int numPoints = randomIntBetween(2, 100);
+        List<Coordinate> coordinates = new ArrayList<>(numPoints);
+        GeoPoint p;
+        for (int i = 0; i < numPoints; ++i) {
+            p = RandomShapeGenerator.randomPointIn(random(), -90d, -90d, 90d, 90d);
+            coordinates.add(new Coordinate(p.lon(), p.lat()));
+        }
+        return coordinates;
+    }
+
+    @Override
+    public void testParseLineString() throws IOException {
+        List<Coordinate> coordinates = randomLineStringCoords();
+        LineString expected = GEOMETRY_FACTORY.createLineString(coordinates.toArray(new Coordinate[coordinates.size()]));
+        assertExpected(jtsGeom(expected), new LineStringBuilder(coordinates));
+    }
+
+    @Override
+    public void testParseMultiLineString() throws IOException {
+        int numLineStrings = randomIntBetween(2, 8);
+        List<LineString> lineStrings = new ArrayList<>(numLineStrings);
+        MultiLineStringBuilder builder = new MultiLineStringBuilder();
+        for (int j = 0; j < numLineStrings; ++j) {
+            List<Coordinate> lsc = randomLineStringCoords();
+            Coordinate [] coords = lsc.toArray(new Coordinate[lsc.size()]);
+            lineStrings.add(GEOMETRY_FACTORY.createLineString(coords));
+            builder.linestring(new LineStringBuilder(lsc));
+        }
+        MultiLineString expected = GEOMETRY_FACTORY.createMultiLineString(
+            lineStrings.toArray(new LineString[lineStrings.size()]));
+        assertExpected(jtsGeom(expected), builder);
+    }
+
+    @Override
+    public void testParsePolygon() throws IOException {
+        PolygonBuilder builder = PolygonBuilder.class.cast(
+            RandomShapeGenerator.createShape(random(), RandomShapeGenerator.ShapeType.POLYGON));
+        Coordinate[] coords = builder.coordinates()[0][0];
+        LinearRing shell = GEOMETRY_FACTORY.createLinearRing(coords);
+        Polygon expected = GEOMETRY_FACTORY.createPolygon(shell, null);
```

Comment:
Should this also work with holes? I just tried parsing the second polygon example from the issue decription and it fails (`POLYGON ((35 10, 45 45, 15 40, 10 20, 35 10), (20 30, 35 35, 30 20, 20 30))`)  I suspect the second argument is a hole so I was trying to see of this was tested, seems like this needs to be added still.

Reply:
:+1:  Good catch! I agree we should have tests for holes. This is a widely used use case. Will add that next.

Reply:
I added an explicit test for polygons with holes below in `testParsePolygonWithHole`, are you asking for a random polygon with random hole? That's a bit more tricky to generate a random polygon with a random inscribed polygon as a hole. Probably not really necessary as long as we've tested at least one polygon with a hole?

Reply:
Yes, I saw the explicit test and I was hoping for a randomized polygon/hole thing, but I just realized that also GeoJsonShapeParserTests doesn't do this fully randomized. So if it is difficult I'm also fine if there is only the explicit test.